addons/account/tests/test_account_move_in_invoice.py

```diff
@@ -2292,3 +2292,41 @@ def test_invoice_sent_to_additional_partner(self):
 
         self.assertIn('access_token=', additional_partner_mail.body_html,
                       "The additional partner should be sent the link including the token")
+
+    def test_onchange_journal_currency(self):
+        """
+        Ensure invoice currency changes on journal change, iff the journal
+        has a currency_id set.
+        """
+
+        chf = self.env.ref('base.CHF')
+        eur = self.env.ref('base.EUR')
+
+        journal_gen_exp = self.env['account.journal'].create({
+            'name': 'Expenses (generic)',
+            'type': 'purchase',
+            'code': 'EXPGEN',
+        })
+        journal_swiss_exp = self.env['account.journal'].create({
+            'name': 'Expenses (CHF)',
+            'type': 'purchase',
+            'code': 'EXPCHF',
+            'currency_id': chf.id,
+        })
+
+        self.invoice.currency_id = eur
+        move_form = Form(self.invoice)
+        invoice = move_form.save()
+        self.assertEqual(invoice.currency_id, eur)
+
+        # No change expected with generic journal
+        move_form.journal_id = journal_gen_exp
+        move_form.save()
+        self.assertEqual(invoice.currency_id, eur,
+                         "Changing to a journal without set currency shouldn't affect invoice currency")
```

Comment:
Unless the company changes; if there is no easy solution I guess it could be ok

Reply:
I'm not sure how that would happen, is there a setting that allows you to select journals from different companies on a vendor bill?

Reply:
By the same token, shouldn't we be handling cases where a user selects a sales journal for a purchase this way?

I'm referring to the expected use case. Using the Accounting or Purchase app, the only journals you can select when creating an invoice are the ones of your current company. I can probably create a fix that handles this case, but it'd add a ton of unneeded complexity to handle an inconvenience that only occurs when using a custom script to modify the database.